src/WidgetBase.ts

```diff
@@ -328,41 +292,57 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 	public __setProperties__(originalProperties: P): void {
 		const { bind, ...properties } = originalProperties as any;
 		const changedPropertyKeys: string[] = [];
-		const allProperties = new Set([...Object.keys(properties), ...Object.keys(this._properties)]);
+		const allProperties = [ ...Object.keys(properties), ...Object.keys(this._properties) ];
+		const checkedProperties: string[] = [];
 		const diffPropertyResults: any = {};
+		const registeredDiffPropertyNames = this.getDecorator('registeredDiffProperty');
+		let runReactions = false;
 
 		this._renderState = WidgetRenderState.PROPERTIES;
-		this._bindFunctionProperties(properties, bind);
 
-		allProperties.forEach((propertyName) => {
+		for (let i = 0; i < allProperties.length; i++) {
+			const propertyName = allProperties[i];
+			if (checkedProperties.indexOf(propertyName) > 0) {
+				continue;
+			}
+			checkedProperties.push(propertyName);
 			const previousProperty = this._properties[propertyName];
-			const newProperty = properties[propertyName];
-			const diffFunctions: DiffPropertyFunction[] = this.getDecorator(`diffProperty:${propertyName}`);
-
-			if (diffFunctions.length === 0) {
-				diffFunctions.push(auto);
+			const newProperty = this._bindFunctionProperty(properties[propertyName], bind);
+			if (registeredDiffPropertyNames.indexOf(propertyName) !== -1) {
+				runReactions = true;
+				const diffFunctions = this.getDecorator(`diffProperty:${propertyName}`);
+				for (let i = 0; i < diffFunctions.length; i++) {
+					const result = diffFunctions[i](previousProperty, newProperty);
+					if (result.changed && changedPropertyKeys.indexOf(propertyName) === -1) {
+						changedPropertyKeys.push(propertyName);
+					}
+					if (propertyName in properties) {
+						diffPropertyResults[propertyName] = result.value;
+					}
+				}
 			}
-
-			diffFunctions.forEach((diffFunction) => {
-				const result = diffFunction.call(null, previousProperty, newProperty);
+			else {
+				const result = boundAuto(previousProperty, newProperty);
 				if (result.changed && changedPropertyKeys.indexOf(propertyName) === -1) {
 					changedPropertyKeys.push(propertyName);
 				}
 				if (propertyName in properties) {
 					diffPropertyResults[propertyName] = result.value;
 				}
-			});
-		});
-
-		this._mapDiffPropertyReactions(properties, changedPropertyKeys).forEach((args, reaction) => {
-			if (args.changed) {
-				reaction.call(this, args.previousProperties, args.newProperties);
 			}
-		});
+		}
+
+		if (runReactions) {
+			this._mapDiffPropertyReactions(properties, changedPropertyKeys).forEach((args, reaction) => {
```

Comment:
speculating, but with us targeting the down emit and access `this` in the arrow function, it might be an opportunity to create a bound callback that we cache if this gets called a lot?